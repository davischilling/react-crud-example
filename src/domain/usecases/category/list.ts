import { api } from '../../api';
import { Category } from '../../models/category';
import { StatefulUseCase } from '../stateful-usecase';

export type State = {
  isLoading: boolean;
  categories: Category[];
  filteredCategories: Category[];
};

export const DEFAULT_STATE: State = {
  isLoading: true,
  categories: [],
  filteredCategories: [],
};

export class ListCategoriesUseCase extends StatefulUseCase<State> {
  public init = async () => {
    await this.initCategories();
  };

  initCategories = async () => {
    const { data } = await api.get('/categories');
    this.loadCategories(data);
  };

  filterCategories = (search: string | undefined) => {
    const categories = this.getState().categories;
    if (!search) {
      this.setState({ filteredCategories: categories });
      return;
    }
    const filteredCategories = categories.filter((category) => {
      return (
        category.name.toLowerCase().includes(search.toLowerCase()) ||
        category.description?.toLowerCase().includes(search.toLowerCase())
      );
    });
    this.setState({ filteredCategories });
  };

  deleteCategory = async (id: string) => {
    const categories = this.getState().categories;
    const category = categories.find((category) => category.id === id);
    if (!category) {
      throw new Error('Category not found');
    }
    const fmtCategory: Category = {
      ...category,
      deleted_at: new Date().toString(),
    };
    await api.put(`/categories/${id}`, fmtCategory);
    const fmrCategories = categories.map((category) => {
      if (category.id === id) {
        return fmtCategory;
      }
      return category;
    });
    this.loadCategories(fmrCategories);
  };

  private loadCategories = (c: Category[], cb?: () => void) => {
    const categories = c.filter((c: Category) => !c.deleted_at);
    this.setState({ categories, filteredCategories: categories }, () => {
      cb && cb();
      this.setState({ isLoading: false });
    });
  };
}
