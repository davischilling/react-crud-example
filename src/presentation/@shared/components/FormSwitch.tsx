import { FormControlLabel, FormGroup, Grid, Switch, SwitchProps } from '@mui/material';

type FormSwitchProps = {
  label: string;
} & SwitchProps;

export default function FormSwitch({ label, ...rest }: FormSwitchProps) {
  return (
    <Grid item xs={12}>
      <FormGroup>
        <FormControlLabel control={<Switch {...rest} />} label={label} />
      </FormGroup>
    </Grid>
  );
}
