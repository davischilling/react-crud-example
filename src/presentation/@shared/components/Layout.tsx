import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import React from 'react';

export function Layout({ children }: { children: React.ReactNode }) {
  return (
    <Box>
      <Container maxWidth="lg" sx={{ mt: 4, mb: 4, color: 'white' }}>
        {children}
      </Container>
    </Box>
  );
}
