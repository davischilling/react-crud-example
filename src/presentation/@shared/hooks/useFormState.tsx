import { useSnackbar } from 'notistack';
import { ChangeEvent, FormEvent, useMemo } from 'react';

type FormStateProps<Name> = {
  setEntity: ({ name, value }: { name: Name; value: string }) => void;
  onToggle?: (value: boolean) => void;
  onSubmit: () => Promise<void>;
  submitCallback: () => void;
  snackMessage: string;
};

export function useFormState<Name = string>({
  setEntity,
  onToggle,
  onSubmit,
  submitCallback,
  snackMessage,
}: FormStateProps<Name>) {
  const { enqueueSnackbar } = useSnackbar();

  const handleChange = useMemo(
    () => (e: ChangeEvent<HTMLInputElement>) => {
      setEntity({
        name: e.target.name as Name,
        value: e.target.value,
      });
    },
    [setEntity],
  );

  const handleToggle = useMemo(
    () => (event: ChangeEvent<HTMLInputElement>, checked: boolean) => {
      onToggle && onToggle(checked);
    },
    [],
  );

  const handleSubmit = useMemo(
    () => async (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      try {
        await onSubmit();
        submitCallback();
        enqueueSnackbar(snackMessage, { variant: 'success' });
      } catch (error) {
        enqueueSnackbar('Something went wrong', { variant: 'error' });
      }
    },
    [],
  );

  return { handleChange, handleToggle, handleSubmit };
}
