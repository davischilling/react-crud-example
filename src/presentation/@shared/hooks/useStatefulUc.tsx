import { useCallback, useEffect, useLayoutEffect, useMemo, useRef, useState } from 'react';
import {
  SetState,
  StateCallback,
  StatefulUseCase,
} from '../../../domain/usecases/stateful-usecase';

type StateFulContextUseCaseProps<State, UseCaseClass extends StatefulUseCase<State>> = {
  UseCase: new (state: State, setStateCb: SetState<State>) => UseCaseClass;
  store: () => [State, (value: Partial<State>) => void];
};

function useCallbackRef<T = () => void>(callback: T) {
  const callbackRef = useRef(callback);
  useLayoutEffect(() => {
    callbackRef.current = callback;
  }, [callback]);
  return callbackRef;
}

export function useStatefulUseCase<State, UseCaseClass extends StatefulUseCase<State>>({
  UseCase,
  store,
}: StateFulContextUseCaseProps<State, UseCaseClass>) {
  const [logicState, setLogicState] = store();

  const setState = useCallback(
    (newState: any, cb?: StateCallback) => {
      setLogicState(Object.assign({}, logicState, newState));
      cb && cb();
    },
    [logicState, setLogicState],
  );

  console.log('rendering useStatefulUseCase');

  const useCaseLogicRef = useCallbackRef<UseCaseClass>(
    new UseCase(logicState, (newState, cb) => {
      setState(newState, cb);
    }),
  );

  useEffect(() => {
    return () => {
      useCaseLogicRef.current.init();
    };
  }, []);

  return { state: logicState, useCase: useCaseLogicRef.current };
}
