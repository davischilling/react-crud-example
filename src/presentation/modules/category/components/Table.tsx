import DeleteIcon from '@mui/icons-material/Delete';
import { Box, IconButton, Typography } from '@mui/material';
import {
  DataGrid,
  GridColDef,
  GridFilterModel,
  GridRenderCellParams,
  GridRowsProp,
  GridToolbar,
} from '@mui/x-data-grid';
import { useSnackbar } from 'notistack';
import { Link } from 'react-router-dom';

import { Category } from '../../../../domain/models/category';

export type TableProps = {
  data: Category[] | undefined;
  isFetching: boolean;
  rowsPerPage: number[];
  handleFilterChange: (filterModel: GridFilterModel) => void;
  handleDelete: (id: string) => Promise<void>;
};

export function Table({
  data,
  handleDelete,
  rowsPerPage,
  handleFilterChange,
  isFetching,
}: TableProps) {
  const componentProps = {
    toolbar: {
      showQuickFilter: true,
      quickFilterProps: { debounceMs: 500 },
    },
  };
  const { enqueueSnackbar } = useSnackbar();

  function mapRows(data: Category[]) {
    return data.map((category) => ({
      id: category.id,
      name: category.name,
      description: category.description,
      is_active: category.is_active,
      created_at: new Date(category.created_at).toLocaleDateString('pt-BR'),
    }));
  }

  const renderLinkToEditPageCell = (row: GridRenderCellParams) => (
    <Link style={{ textDecoration: 'none', color: 'inherit' }} to={`/categories/edit/${row.id}`}>
      <Typography>{row.value}</Typography>
    </Link>
  );

  const renderIsActiveCell = (row: GridRenderCellParams) => (
    <Typography color={row.value ? 'primary' : 'secondary'}>
      {row.value ? 'Active' : 'Inactive'}
    </Typography>
  );
  const renderEditDateCell = (row: GridRenderCellParams) => (
    <Typography data-testid={`created-at-cell-${row.id}`} data-cell-id={row.id}>
      {row.value}
    </Typography>
  );

  const onDelete = async (value: string) => {
    try {
      await handleDelete(value);
      enqueueSnackbar('Category deleted successfully', { variant: 'success' });
    } catch (error) {
      enqueueSnackbar('Error deleting category', { variant: 'error' });
    }
  };

  const renderActionsCell = (row: GridRenderCellParams) => (
    <IconButton
      aria-label="delete"
      onClick={() => onDelete(row.value as string)}
      color="secondary"
      data-testid={`delete-button-${row.id}`}
    >
      <DeleteIcon />
    </IconButton>
  );

  const rows: GridRowsProp = data ? mapRows(data) : [];
  const columns: GridColDef[] = [
    {
      field: 'name',
      headerName: 'Name',
      flex: 1,
      renderCell: renderLinkToEditPageCell,
    },
    { field: 'description', headerName: 'Description', flex: 1 },
    {
      field: 'is_active',
      headerName: 'Active',
      flex: 1,
      type: 'boolean',
      renderCell: renderIsActiveCell,
    },
    {
      field: 'created_at',
      headerName: 'CreatedAt',
      flex: 1,
      renderCell: renderEditDateCell,
    },
    {
      field: 'id',
      headerName: 'Actions',
      flex: 1,
      renderCell: renderActionsCell,
    },
  ];

  return (
    <Box sx={{ display: 'flex', height: 600 }}>
      <DataGrid
        rows={rows}
        columns={columns}
        slotProps={componentProps}
        slots={{ toolbar: GridToolbar }}
        initialState={{
          pagination: { paginationModel: { pageSize: rowsPerPage[0] } },
        }}
        loading={isFetching}
        pageSizeOptions={rowsPerPage}
        onFilterModelChange={handleFilterChange}
        disableColumnFilter={true}
        disableColumnSelector={true}
        disableDensitySelector={true}
        disableRowSelectionOnClick={true}
      />
    </Box>
  );
}
