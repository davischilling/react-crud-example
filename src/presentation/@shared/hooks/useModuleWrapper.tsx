import React from 'react';
import { StatefulUseCase } from '../../../domain/usecases/stateful-usecase';
import createGenericContext from '../contexts/createGenericContext';
import { useStatefulUseCase } from './useStatefulUc';

type ModuleWrapperProps<State, UseCase extends StatefulUseCase<State>> = {
  DEFAULT_STATE: State;
  UseCase: new (
    state: State,
    setStateCb: (value: Partial<State>, cb?: () => void) => void,
  ) => UseCase;
};

export function useModuleWrapper<State, UseCase extends StatefulUseCase<State>>({
  DEFAULT_STATE,
  UseCase,
}: ModuleWrapperProps<State, UseCase>) {
  const { Provider, useStore } = createGenericContext<State>(DEFAULT_STATE);
  const useModuleStore = () => useStore<State>((store) => store);
  const useModuleUseCase = () =>
    useStatefulUseCase<State, UseCase>({
      UseCase,
      store: useModuleStore,
    });
  return { Provider, useModuleUseCase };
}
