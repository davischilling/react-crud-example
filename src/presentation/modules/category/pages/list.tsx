import { Box, Button, Typography } from '@mui/material';
import { GridFilterModel } from '@mui/x-data-grid';
import { useMemo } from 'react';
import { Link } from 'react-router-dom';

import {
  DEFAULT_STATE,
  ListCategoriesUseCase,
  State,
} from '../../../../domain/usecases/category/list';
import { useStatefulUseCase } from '../../../@shared/hooks/useStatefulUc';
import { Table } from '../components/Table';

export function ListCategoryPage() {
  const { state, useCase } = useStatefulUseCase<State, ListCategoriesUseCase>({
    UseCase: ListCategoriesUseCase,
    DEFAULT_STATE,
  });

  return (
    <Box maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
      <Box p={2}>
        <Box mb={2}>
          <Typography variant="h4">List Categories</Typography>
        </Box>
      </Box>
      <Box display="flex" justifyContent="flex-end">
        <Button
          variant="contained"
          color="secondary"
          component={Link}
          to="/categories/create"
          style={{ marginBottom: '1rem' }}
        >
          New Category
        </Button>
      </Box>

      <Table
        data={state.filteredCategories}
        isFetching={state.isLoading}
        rowsPerPage={[5, 10, 15, 20]}
        handleDelete={useMemo(() => useCase.deleteCategory, [])}
        handleFilterChange={useMemo(
          () => (filterModel: GridFilterModel) =>
            useCase.filterCategories(
              filterModel.quickFilterValues && filterModel.quickFilterValues[0],
            ),
          [],
        )}
      />
    </Box>
  );
}
