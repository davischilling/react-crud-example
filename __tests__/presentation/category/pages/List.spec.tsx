import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import {
  DEFAULT_STATE,
  ListCategoriesUseCase,
} from '../../../../src/domain/usecases/category/list';
import * as useStatefulUseCaseModule from '../../../../src/presentation/@shared/hooks/useStatefulUc';

import { Table, TableProps } from '../../../../src/presentation/modules/category/components/Table';
import { ListCategoryPage } from '../../../../src/presentation/modules/category/pages/list';

jest.mock('../../../../src/presentation/modules/category/components/Table', () => ({
  Table: jest.fn(({ data, isFetching, handleFilterChange }: TableProps) => (
    <div data-testid="mocked-table">
      Mocked Table Content
      <div>
        {!!isFetching ? (
          <>
            <p>Loading...</p>
          </>
        ) : (
          <>
            {data?.length
              ? data.map((el) => (
                  <p key={el.id} data-testid={`cell-id-${el.id}`}>
                    {el.name}
                  </p>
                ))
              : 'No rows'}
          </>
        )}
        <button
          onClick={() =>
            handleFilterChange({
              quickFilterValues: ['Category 1'],
              items: [],
            })
          }
        >
          filter
        </button>
      </div>
    </div>
  )),
}));

describe('ListCategoryPage', () => {
  let fragment: DocumentFragment;
  let useStatefulUseCase: jest.SpyInstance;
  let filterCategoriesSpy: jest.SpyInstance;
  const mockedData = [
    { id: 1, name: 'Category 1' },
    { id: 2, name: 'Category 2' },
  ];

  beforeEach(async () => {
    jest.clearAllMocks();
    filterCategoriesSpy = jest.fn();
    useStatefulUseCase = jest
      .spyOn(useStatefulUseCaseModule as any, 'useStatefulUseCase')
      .mockReturnValue({
        state: {
          isLoading: false,
          filteredCategories: mockedData,
        },
        useCase: {
          filterCategories: filterCategoriesSpy,
        },
      });
    fragment = await waitFor(() => {
      const { asFragment } = render(<ListCategoryPage />, { wrapper: BrowserRouter });
      return asFragment();
    });
  });

  test('renders the "List Categories" page', () => {
    expect(fragment).toMatchSnapshot();
  });

  test('calls useStatefulUseCase with correct arguments', () => {
    expect(useStatefulUseCase).toHaveBeenCalledTimes(1);
    expect(useStatefulUseCase).toHaveBeenCalledWith({
      UseCase: ListCategoriesUseCase,
      DEFAULT_STATE,
    });
  });

  test('renders the "New Category" button', () => {
    const newCategoryButton = screen.getByText('New Category');
    expect(newCategoryButton).toBeInTheDocument();
  });

  test('renders the grid component', async () => {
    const grid = screen.getByTestId('mocked-table');
    expect(grid).toBeInTheDocument();
  });

  test('renders the Table component with correct data', async () => {
    expect(Table).toHaveBeenCalledWith(
      expect.objectContaining({
        data: mockedData,
        isFetching: false,
        rowsPerPage: [5, 10, 15, 20],
      }),
      expect.any(Object),
    );
  });

  test('calls useStatefulUseCase with correct arguments', () => {
    expect(useStatefulUseCase).toHaveBeenCalledTimes(1);
    expect(useStatefulUseCase).toHaveBeenCalledWith({
      UseCase: ListCategoriesUseCase,
      DEFAULT_STATE,
    });
  });

  test('renders the "New Category" button', () => {
    const newCategoryButton = screen.getByText('New Category');
    expect(newCategoryButton).toBeInTheDocument();
  });

  test('renders the grid component', async () => {
    const grid = screen.getByTestId('mocked-table');
    expect(grid).toBeInTheDocument();
  });

  test('renders the Table component with correct data when filtered categories are not empty', async () => {
    expect(Table).toHaveBeenCalledWith(
      expect.objectContaining({
        data: mockedData,
        isFetching: false,
        rowsPerPage: [5, 10, 15, 20],
      }),
      expect.any(Object),
    );
  });

  test('renders the component with loading state', async () => {
    useStatefulUseCase.mockReturnValue({
      state: {
        isLoading: true,
        filteredCategories: [],
      },
      useCase: {},
    });

    const screen = await waitFor(() => render(<ListCategoryPage />, { wrapper: BrowserRouter }));

    const loadingIndicator = screen.getByText('Loading...');
    expect(loadingIndicator).toBeInTheDocument();
  });

  test('renders the component with non-empty filtered categories', async () => {
    jest.clearAllMocks();
    useStatefulUseCase.mockReturnValue({
      state: {
        isLoading: false,
        filteredCategories: mockedData,
      },
      useCase: {},
    });

    const screen = await waitFor(() => render(<ListCategoryPage />, { wrapper: BrowserRouter }));

    const el1 = screen.getAllByTestId('cell-id-1')[0];
    const el2 = screen.getAllByTestId('cell-id-2')[0];
    const el3 = screen.queryByTestId('cell-id-3');

    expect(el1).toBeInTheDocument();
    expect(el2).toBeInTheDocument();
    expect(el3).not.toBeInTheDocument();
  });

  test('renders the component with empty filtered categories', async () => {
    useStatefulUseCase.mockReturnValueOnce({
      state: {
        isLoading: false,
        filteredCategories: [],
      },
      useCase: {},
    });

    await waitFor(() => render(<ListCategoryPage />, { wrapper: BrowserRouter }));
    const noDataMessage = screen.getByText('No rows');
    expect(noDataMessage).toBeInTheDocument();
  });

  test('calls filterCategories with correct arguments', async () => {
    const filterBtn = screen.getByText('filter');

    await waitFor(() => {
      fireEvent.click(filterBtn);
    });
    expect(filterCategoriesSpy).toHaveBeenCalledTimes(1);
    expect(filterCategoriesSpy).toHaveBeenCalledWith('Category 1');
  });
});
