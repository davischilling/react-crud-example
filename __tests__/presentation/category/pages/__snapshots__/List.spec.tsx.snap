// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListCategoryPage renders the "List Categories" page 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1ozyran"
  >
    <div
      class="MuiBox-root css-19midj6"
    >
      <div
        class="MuiBox-root css-1qm1lh"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          List Categories
        </h4>
      </div>
    </div>
    <div
      class="MuiBox-root css-1bvc4cc"
    >
      <a
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        href="/categories/create"
        style="margin-bottom: 1rem;"
        tabindex="0"
      >
        New Category
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
    </div>
    <div
      data-testid="mocked-table"
    >
      Mocked Table Content
      <div>
        <p
          data-testid="cell-id-1"
        >
          Category 1
        </p>
        <p
          data-testid="cell-id-2"
        >
          Category 2
        </p>
        <button>
          filter
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
